src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+
+        uint256 tic = ilks[ilk].tic;
+        require(tic > 0, "DssDirectDepositHub/pool-live");
+
+        uint256 tau = ilks[ilk].tau;
+        require(_add(tic, tau) <= block.timestamp || wards[msg.sender] == 1, "DssDirectDepositHub/unauthorized-cull");
+
+        uint256 culled = ilks[ilk].culled;
+        require(culled == 0, "DssDirectDepositHub/already-culled");
+
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(pool));
+        require(ink <= 2 ** 255, "DssDirectDepositHub/overflow");
+        require(art <= 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.grab(ilk, address(pool), address(pool), vow, -int256(ink), -int256(art));
+
+        if (ink > art) {
+            // We have more collateral than debt, so need to rebalance.
+            // After cull the gems we grab above represent the debt to
+            // unwind.
+            vat.slip(ilk, address(pool), -int256(ink - art));
+        }
+
+        ilks[ilk].culled = 1;
+        emit Cull(ilk);
+    }
+
+    /**
+        @notice Rollback Write-off (cull) if General Shutdown happened.
+        This function is required to have the collateral back in the vault so it
+        can be taken by End module and eventually be shared to DAI holders (as
+        any other collateral) or maybe even unwinded.
+        @dev This pulls gems from the pool and reopens the urn with the gem
+        amount of ink/art.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function uncull(bytes32 ilk) external {
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        require(ilks[ilk].culled == 1, "DssDirectDepositHub/not-prev-culled");
+        require(vat.live() == 0, "DssDirectDepositHub/no-uncull-normal-operation");
+
+        address vow_ = vow;
+        uint256 wad = vat.gem(ilk, address(pool));
+        require(wad < 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.suck(vow_, vow_, _mul(wad, RAY)); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+        vat.grab(ilk, address(pool), address(pool), vow_, int256(wad), int256(wad));
+
+        ilks[ilk].culled = 0;
+        emit Uncull(ilk);
+    }
+
+    /**
+        @notice Emergency Quit Everything.
+        Transfer all the shares and either wipe out the gems (culled situation)
+        or fork the urn to the recipient.
+        @dev If called while not culled, it will require who to hope on the Hub
+        contract in the Vat.
+        @param ilk bytes32 of the D3M ilk name
+        @param who address of who will receive the shares and possibly the urn
+    */
+    function quit(bytes32 ilk, address who) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-quit-during-shutdown");
+
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        // Send all gem in the contract to who
+        require(pool.transferAll(who), "DssDirectDepositHub/failed-transfer");
```

Comment:
Potential reentrancy issue having this transfer before system accounting. 
Move this to the end of the execution if possible to follow checks-effects-interactions pattern.

Quit could also benefit from a reentrancy guard.

Reply:
09cc6c4